File: BFS/847.Shortest-Path-Visiting-All-Nodes/847.Shortest-Path-Visiting-All-Nodes.py
class Solution:
    def shortestPathLength(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: int
        """
        N = len(graph)
        allVisitedStatus = (1<<N)-1
        
        q = collections.deque()
        visited = set()
        for i in range(N):
            q.append((i,1<<i))
            visited.add((i,1<<i))
        
        step = 0
        while (q):
            L = len(q)
            step += 1
            for i in range(L):
                node,status = q.popleft()
                for nextNode in graph[node]:         
                    if status|(1<<nextNode)==allVisitedStatus:
                        return step
                    if (nextNode,status|(1<<nextNode)) in visited: 
                        continue
                    q.append((nextNode,status|(1<<nextNode)))
                    visited.add((nextNode,status|(1<<nextNode)))        
        return 0

File: BFS/847.Shortest-Path-Visiting-All-Nodes/847.Shortest-Path-Visiting-All-Nodes.cpp
class Solution {
    int visited[12][1<<12];
public:
    int shortestPathLength(vector<vector<int>>& graph) 
    {
        int n=graph.size();        
                       
        queue<pair<int,int>>q;        
                
        for (int i=0; i<n; i++)
        {            
            q.push({i,1<<i});
            visited[i][1<<i] = 1;
        }
        
        int step = 0;
        while (q.size()!=0)
        {
            int len = q.size();
            while (len--)
            {
                auto [cur, state] = q.front();
                q.pop();
                
                for (int i: graph[cur])
                {                    
                    if ((state | (1<<i)) == (1<<n)-1) return step+1;                        
                    if (visited[i][state]==1) continue;
                    visited[i][state] = 1;
                    q.push({i, state | (1<<i)});
                }                
            }
            step++;            
        }
        
        return 0;
    }

};
